File: AddTwoNumbers/AddTwoNumbers.py
/*
You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero, except the number 0 itself.
*/

/*
Runtime: 76 ms, faster than 85.42% of Python3 online submissions for Add Two Numbers.
Memory Usage: 13.1 MB, less than 85.25% of Python3 online submissions for Add Two Numbers.
*/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def calcVal(self, l: ListNode) -> int:
        if l:
            return l.val
        return 0
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummyhead = ListNode(0)
        cur = dummyhead
        carry = False
        while l1 or l2 or carry:
          sum = self.calcVal(l1) + self.calcVal(l2) + carry
          next = sum % 10
          carry = sum // 10
          cur.next = ListNode(next)
          cur = cur.next
          if l1:
            l1 = l1.next
          if l2:
            l2 = l2.next
        return dummyhead.next

File: AddTwoNumbers/AddTwoNumbers.cpp
/*
You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order
and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero, except the number 0 itself.
*/

/*
Runtime: 24 ms, faster than 84.78% of C++ online submissions for Add Two Numbers.
Memory Usage: 10.4 MB, less than 54.28% of C++ online submissions for Add Two Numbers.
*/

/*
struct ListNode {
      int val;
      ListNode *next;
      ListNode(int x) : val(x), next(NULL) {}
  };
*/
    
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        bool plus = false;
        int sum;
        ListNode *beforetop = new ListNode(0);        
        ListNode *result = beforetop;
        while(l1 || l2 || plus) {
            int sum = calcVal(l1) + calcVal(l2) + plus;
            result->next = new ListNode(sum % 10);
            plus = sum / 10;
            result = result->next;
            
            if(l1) {
                l1 = l1->next;
            }
            if(l2) {
                l2 = l2->next;
            }
        }
        return beforetop->next;
    }
    int calcVal(ListNode * lp) {
        if(lp) {
            return lp->val;
        }
        return 0;
    }
};
